### PR TITLE
feat: Add Voting watcher bot

### PR DESCRIPTION
**Pull request by dgusakov**

### CODE REVIEW

File: voting-watcher/src/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    seconds %= 3600 * 24;
+    timeStrings.push(`${days.toFixed(0)} ${days > 1 ? "days" : "day"}`);
+  }
+  const hours = Math.floor(seconds / 3600);
+  if (hours > 0) {
+    timeStrings.push(`${hours.toFixed(0)} ${hours > 1 ? "hours" : "hour"}`);
+  }
+  seconds %= 3600;
+  const minutes = Math.floor(seconds / 60);
+  if (timeStrings.length == 1 && minutes > 0) {
```
Comment (madlabman) [inline voting-watcher/src/helpers.ts:48]: I didn't get the idea here. If we have minutes in (0;60) we won't print anything?
What is the case for that?

File: voting-watcher/src/agent-voting-watcher.ts [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+  const prevBlock = await ethersProvider.getBlock(blockEvent.blockNumber - 1);
+  Array.from(activeVotes.keys()).forEach((key) => {
+    const vote = activeVotes.get(key);
+    if (vote && !vote.passed) {
+      PINGER_SCHEDULE.forEach((time) => {
+        const pingTime =
+          vote.startDate + voteLength - objectionsTime - time * 3600;
+        if (
+          pingTime <= blockEvent.block.timestamp &&
+          pingTime > prevBlock.timestamp
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts:162]: To ping only once

File: voting-watcher/src/helpers.ts (hunk 1 above)
Comment (dgusakov) [reply voting-watcher/src/helpers.ts:48]: We need it not to print "0 minutes". And yes we drop seconds here

File: voting-watcher/src/agent-voting-watcher.ts [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    ...Array.from(newActiveVotes.keys()),
+  ]).forEach((key) => {
+    const updated = newActiveVotes.get(key);
+    const old = activeVotes.get(key);
+    if (updated && old) {
+      const { url, pro, contra, resultsStr, timeLeft } = updated;
+      const proStr = pro.toFixed(2) + "% pro";
+      const contraStr = contra.toFixed(2) + "% contra";
+      const alertLevel = old.alertLevel || 0;
+      if (alertLevel < 2 && timeLeft < 3600 * 4 && updated.phase == 1) {
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts:97]: It is not. It guarantees that alert will fire only once

[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts:97]: Mess with phase values fixed

File: voting-watcher/src/agent-voting-watcher.ts [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+  return findings;
+}
+
+function handleAragonEventsOfNoticeTransaction(
+  txEvent: TransactionEvent,
+  findings: Finding[]
+) {
+  ARAGON_VOTING_EVENTS_OF_NOTICE.forEach((eventInfo) => {
+    if (eventInfo.address in txEvent.addresses) {
+      const [event] = txEvent.filterLog(eventInfo.event, eventInfo.address);
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts:185]: Fixed

File: voting-watcher/src/agent-voting-watcher.ts [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+      name: "Vote state changed",
+      description: text,
+      alertId: "VOTE-STATE-CHANGED",
+      severity: FindingSeverity.Info,
+      type: FindingType.Info,
+    })
+  );
+}
+
+function hugeVotes(
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts:303]: Fixed

File: voting-watcher/src/agent-voting-watcher.ts [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+      const proStr = pro.toFixed(2) + "% pro";
+      const contraStr = contra.toFixed(2) + "% contra";
+      const text =
+        `${formatLink(`Voting #${key}`, url)} is over 🏁 ${
+          passed ? "and passed, phew! 👍" : "but rejected 😔"
+        } ` + `(${proStr}, ${contraStr}, ${resultsStr})`;
+      voteStateChanged(text, findings);
+      activeVotes.delete(key);
+    } else {
+      activeVotes.delete(key);
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts:126]: Fixed

File: voting-watcher/package.json [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  },
+  "devDependencies": {
+    "@types/jest": "^27.0.1",
+    "@types/nodemon": "^1.19.0",
+    "jest": "^27.0.6",
+    "nodemon": "^2.0.8",
+    "ts-jest": "^27.0.3",
+    "typescript": "^4.3.4"
+  },
+  "packageManager": "yarn@2.4.3"
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/package.json:37]: Done

File: voting-watcher/src/agent-voting-watcher.ts [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+    VOTING_ABI,
+    ethersProvider
+  );
+
+  voteLength = new BigNumber(
+    String(await voting.functions.voteTime(blockTag))
+  ).toNumber();
+  objectionsTime = new BigNumber(
+    String(await voting.functions.objectionPhaseTime(blockTag))
+  ).toNumber();
```
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts:215]: Can't see why it is required

[the remark this answers is not in the document]
Comment (madlabman) [reply voting-watcher/src/agent-voting-watcher.ts:218]: It's a problem with state-sensitive tests when a module tracks its private variables.

[the remark this answers is not in the document]
Comment (madlabman) [reply voting-watcher/src/agent-voting-watcher.ts:218]: By the way, you don't have any tests written for the agent.

File: voting-watcher/src/helpers.ts (hunk 1 above)
Comment (madlabman) [reply voting-watcher/src/helpers.ts:48]: I mean there is a flaw:
```
(3600) => 1 hour
(3660) => 1 hour 1 minute
(300) => less than a minute - which is definitely wrong
```

File: voting-watcher/src/agent-voting-watcher.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (madlabman) [reply voting-watcher/src/agent-voting-watcher.ts]: What do all these numbers stand for? That's the reason we have these troubles.

**Comment by madlabman** [comment]
Please add at least an E2E test.

File: voting-watcher/src/agent-voting-watcher.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts]: `alertLevel`- a guarantee of the "only once" alert
`phase` - phase number required to define if phase was switched

Pls review this piece of code one more time and return with the exact issues if any

File: voting-watcher/src/helpers.ts (hunk 1 above)
Comment (dgusakov) [reply voting-watcher/src/helpers.ts]: Nice catch! Fixed + tests added

File: voting-watcher/src/agent-voting-watcher.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (dgusakov) [reply voting-watcher/src/agent-voting-watcher.ts:218]: This agent is for internal use only and has pretty specific logic. No need for tests ATM

**Review by madlabman — APPROVED** [review]

**Pull request merged**